src/cluster.c

```diff
@@ -1496,5982 +522,848 @@ int verifyClusterNodeId(const char *name, int length) {
-     * node is a slave and the request is about a hash slot our master
-     * is serving, we can reply without redirection. */
-    int is_write_command = (cmd_flags & CMD_WRITE) ||
-                           (c->cmd->proc == execCommand && (c->mstate.cmd_flags & CMD_WRITE));
-    if (((c->flags & CLIENT_READONLY) || is_pubsubshard) &&
-        !is_write_command &&
-        nodeIsSlave(myself) &&
-        myself->slaveof == n)
-    {
-        return myself;
-    }
-
-    /* Base case: just return the right node. However if this node is not
-     * myself, set error_code to MOVED since we need to issue a redirection. */
-    if (n != myself && error_code) *error_code = CLUSTER_REDIR_MOVED;
-    return n;
-}
-
-/* Send the client the right redirection code, according to error_code
- * that should be set to one of CLUSTER_REDIR_* macros.
- *
- * If CLUSTER_REDIR_ASK or CLUSTER_REDIR_MOVED error codes
- * are used, then the node 'n' should not be NULL, but should be the
- * node we want to mention in the redirection. Moreover hashslot should
- * be set to the hash slot that caused the redirection. */
-void clusterRedirectClient(client *c, clusterNode *n, int hashslot, int error_code) {
-    if (error_code == CLUSTER_REDIR_CROSS_SLOT) {
-        addReplyError(c,"-CROSSSLOT Keys in request don't hash to the same slot");
-    } else if (error_code == CLUSTER_REDIR_UNSTABLE) {
-        /* The request spawns multiple keys in the same slot,
-         * but the slot is not "stable" currently as there is
-         * a migration or import in progress. */
-        addReplyError(c,"-TRYAGAIN Multiple keys request during rehashing of slot");
-    } else if (error_code == CLUSTER_REDIR_DOWN_STATE) {
-        addReplyError(c,"-CLUSTERDOWN The cluster is down");
-    } else if (error_code == CLUSTER_REDIR_DOWN_RO_STATE) {
-        addReplyError(c,"-CLUSTERDOWN The cluster is down and only accepts read commands");
-    } else if (error_code == CLUSTER_REDIR_DOWN_UNBOUND) {
-        addReplyError(c,"-CLUSTERDOWN Hash slot not served");
-    } else if (error_code == CLUSTER_REDIR_MOVED ||
-               error_code == CLUSTER_REDIR_ASK)
-    {
-        /* Redirect to IP:port. Include plaintext port if cluster is TLS but
-         * client is non-TLS. */
-        int use_pport = (server.tls_cluster &&
-                        c->conn && (c->conn->type != connectionTypeTls()));
-        int port = use_pport && n->pport ? n->pport : n->port;
-        addReplyErrorSds(c,sdscatprintf(sdsempty(),
-            "-%s %d %s:%d",
-            (error_code == CLUSTER_REDIR_ASK) ? "ASK" : "MOVED",
-            hashslot, getPreferredEndpoint(n), port));
-    } else {
-        serverPanic("getNodeByQuery() unknown error.");
-    }
-}
-
-/* This function is called by the function processing clients incrementally
- * to detect timeouts, in order to handle the following case:
- *
- * 1) A client blocks with BLPOP or similar blocking operation.
- * 2) The master migrates the hash slo
```

Comment:
We could drop clusterNodeIsMyself() if we fix the module.c and then also change getReplOffset() to also take the current replicationGetSlaveOffset().

Reply:
clusterNodeIsMyself is currently in use in module.c as well. I think it is worth leaving in place the bits required to maintain naive support of the current module.c stuff and then deprecating the whole concept and at that point removing the parts of the cluster API that are not longer required.

Reply:
Ack